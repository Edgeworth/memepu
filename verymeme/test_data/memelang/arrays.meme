// status: ok
// input: ''
// output: ''
// returns: 4
fn takesbyvalue(arr: [42]i8) i8 {
  ret arr[0];
}

fn takesarrbyptr(arr: *[42]i8) i8 {
  ret arr[0];
}

fn takesbyptr(arr: *i8) i8 {
  ret arr[0];
}

fn takesbyref(arr: &[42]i8) i8 {
  ret arr[0];
}

fn main() i8 {
  arr := [42]i8{0, 1, 2, 3};
  takesbyvalue(arr);
  takesarrbyptr(&arr);
  takesbyref(arr);
  takesbyptr(&arr[0]);
  ret 4;
}
