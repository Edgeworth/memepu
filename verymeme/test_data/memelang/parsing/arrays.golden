File
 Fn
  FnSig
   Typename(takesbyvalue)
   VarDecl
    VarRef(arr)
    Type(i8)
     Qualifier(const: 0 ptr: 0, array: IntLit(42))
      IntLit(42)
   Type(i8)
  StmtBlk
   Return
    Op(ARRAY_ACCESS)
     VarRef(arr)
     IntLit(0)
 Fn
  FnSig
   Typename(takesbyref)
   VarDecl
    VarRef(arr)
    Type(i8)
     Qualifier(const: 0 ptr: 1, array: 0)
     Qualifier(const: 0 ptr: 0, array: IntLit(42))
      IntLit(42)
   Type(i8)
  StmtBlk
   Return
    Op(ARRAY_ACCESS)
     VarRef(arr)
     IntLit(0)
 Fn
  FnSig
   Typename(takesbyptr)
   VarDecl
    VarRef(arr)
    Type(i8)
     Qualifier(const: 0 ptr: 1, array: 0)
   Type(i8)
  StmtBlk
   Return
    Op(ARRAY_ACCESS)
     VarRef(arr)
     IntLit(0)
 Fn
  FnSig
   Typename(main)
   Type(i8)
  StmtBlk
   VarDefn
    VarDecl
     VarRef(arr)
     Type(i8)
      Qualifier(const: 0 ptr: 0, array: IntLit(42))
       IntLit(42)
    CompoundLit
   Op(FN_CALL)
    VarRef(takesbyvalue)
    FnCallArgs
     VarRef(arr)
   Op(FN_CALL)
    VarRef(takesbyref)
    FnCallArgs
     Op(UNARY_ADDR)
      VarRef(arr)
   Op(FN_CALL)
    VarRef(takesbyptr)
    FnCallArgs
     Op(UNARY_ADDR)
      VarRef(arr)
   Return
    IntLit(4)
 Enum(0 untyped)
  Typename(Unit)
